source/slang/type-layout.h

```diff
@@ -36,21 +36,161 @@ enum class LayoutRulesFamily
 };
 #endif
 
+// A "size" that can either be a simple finite size or
+// the special case of an infinite/unbounded size.
+//
+struct LayoutSize
+{
+    typedef size_t RawValue;
+
+    LayoutSize()
+        : raw(0)
+    {}
+
+    LayoutSize(RawValue size)
+        : raw(size)
+    {
+        SLANG_ASSERT(size != RawValue(-1));
```

Comment:
Might want to define the 'infinite' constant value somewhere to test against, rather than use the magic RawValue(-1).